### PR TITLE
Fix `logpdf_grad` for BroadcastedNormal.

### LINKED ISSUE #432: `logpdf_grad(::BroadcastedNormal, ...)` sums over the wrong dims
> the `logpdf_grad` implementation of `broadcasted_norm` seems to only handle the cases where _all_ arguments are scalars (note that it returns 3 scalars as gradients)

Good catch, thanks @MrVPlusOne!  Indeed, in https://github.com/probcomp/Gen.jl/commit/005f97c7176ba19cd6f6b88d00cfa2cb32f25c92#, I blindly inserted a `sum` into both `logpdf` and `logpdf_grad`... whoops.  And indeed, what `logpdf_grad` should instead do is sum over the dims that are "missing" from each arg, compared to the fully broadcasted shape, as you do above.

@MrVPlusOne, if you're up for sending a PR with this fix, I'd be happy to review it!

_Originally posted by @bzinberg in https://github.com/probcomp/Gen.jl/issues/420#issuecomment-879447064_

### PR DESCRIPTION
**Pull request by MrVPlusOne**
As requested by #432, this PR fixes the gradient of BroadcastedNormal for the non-scalar cases.

### CODE REVIEW

**Comment by ztangent** [comment]
Thanks for making this PR! I'm not sure why the Travis build errored, but separate of that, it'd great if you could adjust the implementation to be overflow-safe in line with #321!

File: src/modeling_library/distributions/normal.jl [hunk 1]
```diff
@@ -85,10 +83,29 @@ function logpdf_grad(::BroadcastedNormal,
     assert_has_shape(x, broadcast_shapes_or_crash(mu, std);
                      msg="Shape of `x` does not agree with the sample space")
     z = (x .- mu) ./ std
-    deriv_x = sum(- z ./ std)
+    deriv_x = - z ./ std
     deriv_mu = -deriv_x
-    deriv_std = sum(-1. ./ std .+ abs2.(z) ./ std)
-    (deriv_x, deriv_mu, deriv_std)
+    deriv_std = -1. ./ std .+ abs2(z) ./ std
```
Comment (bzinberg) [inline src/modeling_library/distributions/normal.jl]: We need to keep `z`'s shape, right?
```suggestion
    deriv_std = -1. ./ std .+ abs2.(z) ./ std
```

File: src/modeling_library/distributions/normal.jl [hunk 2]
```diff
@@ -85,10 +83,29 @@ function logpdf_grad(::BroadcastedNormal,
...
+    (unbroadcast_for_arg(x, deriv_x), 
+    unbroadcast_for_arg(mu, deriv_mu), 
+    unbroadcast_for_arg(std, deriv_std))
```
Comment (bzinberg) [inline src/modeling_library/distributions/normal.jl]:
```suggestion
    (unbroadcast_for_arg(x, deriv_x), 
     unbroadcast_for_arg(mu, deriv_mu), 
     unbroadcast_for_arg(std, deriv_std))
```

File: test/modeling_library/distributions.jl [hunk 3]
```diff
@@ -126,9 +126,9 @@ end
     f = (x, mu, std) -> logpdf(broadcasted_normal, x, mu, std)
     args = (fill(0.4), fill(0.2), fill(0.3))
     actual = logpdf_grad(broadcasted_normal, args...)
-    @test isapprox(actual[1], finite_diff(f, args, 1, dx; broadcast=true))
-    @test isapprox(actual[2], finite_diff(f, args, 2, dx; broadcast=true))
-    @test isapprox(actual[3], finite_diff(f, args, 3, dx; broadcast=true))
```
Comment (bzinberg) [inline test/modeling_library/distributions.jl]: Why this change?

**Review by bzinberg — COMMENTED** [review]
Thanks @MrVPlusOne!  A couple of initial comments now; I'll come back later to suggest unit / regression tests.

File: src/modeling_library/distributions/normal.jl (hunk 1 above)
Comment (MrVPlusOne) [reply src/modeling_library/distributions/normal.jl]: Yeah, you are absolutely right!

File: test/modeling_library/distributions.jl (hunk 3 above)
Comment (MrVPlusOne) [reply test/modeling_library/distributions.jl]: I'm actually a little unsure how to fix these tests. It seems that the first argument of `isapprox` is a zero-dimensional array (which I believe is the expected behavior) but the second argument is a `Float64`, which causes Julia to complain that no method of `isapprox` matches the given argument types. Any suggestions on how to fix this?

Comment (bzinberg) [reply test/modeling_library/distributions.jl]: After investigating this a bit, I think it boils down to https://github.com/JuliaLang/julia/issues/28866, which I consider to be a bug in Julia.  Yes -- in the version of the code before this PR, both the LHS and the RHS should be 0-dimensional arrays.  For the RHS, the issue appears to be that these lines
https://github.com/probcomp/Gen.jl/blob/d79aded6909a8e44b3a813179758add8237ed46a/test/runtests.jl#L22-L23
do `::Aray{T,0} .+ ::T` and `::Aray{T,0} .- ::T`, which should give an `::Array{T,0}` but due to https://github.com/JuliaLang/julia/issues/28866 gives a scalar.  So the LHS of the `./` on the following line should have an `Array{T,0}` as the `i`th argument to `f`, thus (since `broadcast = true`) should output an `Array{T,0}`, and consequently the division operation should be `::Array{T,0} ./ ::T` which should output an `Array{T,0}`, but that is not what happens.

I think the cleanest way to fix this is to add a workaround to the (test-only) function `finite_diff` so that it handles zero-dimensional array valued arguments correctly.

Comment (bzinberg) [reply test/modeling_library/distributions.jl:131]: Oh whoops, I take that back -- both sides of the `./` are scalars:
https://github.com/probcomp/Gen.jl/blob/d79aded6909a8e44b3a813179758add8237ed46a/test/runtests.jl#L24
I guess what my previous comment really shows is that, if `f` requires its argument to be a zero-dimensional array and errors if given a scalar, `finite_diff` will hit that error.

A fix to `finite_diff` for the zero-dimensional case still seems right :slightly_smiling_face:

Comment (bzinberg) [reply test/modeling_library/distributions.jl]: (I'm currently drafting this.)

Comment (bzinberg) [reply test/modeling_library/distributions.jl]: Aha, in addition to the above, the function `f` being supplied to `finite_diff` does not satisfy the requirements in its docstring: it takes array-valued arguments, but is not the broadcast of a function that takes scalar-valued arguments (i.e. `logpdf_grad(vector-valued distribution parameters)` is not a broadcast of a bunch of `logpdf_grad(scalar-valued distribution parameters)`'s).  So I guess we should add a variant of `finite_diff` that has the semantics we need.

Comment (MrVPlusOne) [reply test/modeling_library/distributions.jl]: Yeah, I think to properly test this, we will need a more general `finite_diff` function or use an autodiff library implementation as a reference.

Comment (bzinberg) [reply test/modeling_library/distributions.jl]: you clearly trust autodiff more than I do :slightly_smiling_face:

**Comment by bzinberg** [comment]
@MrVPlusOne, I've drafted a fix to the unit test breakage.  Take a look and LMKWYT?

**Comment by MrVPlusOne** [comment]
@bzinberg Thanks! The fix looks good to me!

**Comment by bzinberg** [comment]
@MrVPlusOne, I've made a few more changes that are mostly by way of tidying.

* Rename the vars in the `unbroadcast` helper functions to generically describe what they do, rather than coupling them to how they're used in `logpdf_grad`; e.g., `unbroadcast_for_arg(arg, grad)` becomes `_unbroadcast_like(a, full_arr)`
* Add some documentation, mainly for what "unbroadcast" means since I can imagine that being ambiguous or confusing to a reader who doesn't have the same context we have
* A bit of code tidying without changing the logic
* Line length and indentation

The one functional change I made was to add to a couple of the unit tests an explicit check on the shape of the returned arrays.

I think regression testing is covered by the shape checks that were fixed in https://github.com/probcomp/Gen.jl/pull/433/commits/e712c71b56f9b86b3d6bd872cfcc1c41052f8f99.  They passed before, but that is because the unit tests were incorrect.

@MrVPlusOne, any comments you'd like to make on the above?  If looks good to you, I think this is ready to merge.

**Comment by MrVPlusOne** [comment]
Thanks for the cleaning up; they all look good to me! I think one more nice thing to have would be to add a unit test for the case where the two arguments of `broadcasted_norm` have different numbers of dimensions. We currently don't have a test for that, right?

**Comment by bzinberg** [comment]
Great idea @MrVPlusOne, how's [this](https://github.com/probcomp/Gen.jl/pull/433/commits/5d3f192dc03069956f7b4e23e6889e84857aa744)?

**Comment by MrVPlusOne** [comment]
Thanks, looks good to me!

**Comment by bzinberg** [comment]
Thanks @MrVPlusOne for using and improving Gen!
